EventFilter.error mutes all errors regardless of the cause, #25057

dotta opened the pull request.

dotta committed:
EventFilter.error filters all errors regardless of the cause, #25057

ktoso commented:
resolves #25057

ktoso commented:
OK TO TEST

ktoso commented:
PLS WHITELIST

dotta commented:
Ah, I just realized I should amend the commit message and change "mutes" with "filter". Shall I make the change and force push or wait for the review process to complete first?

patriknw reviewed (approved):
LGTM, thanks @dotta

ktoso commented:
Commit message is not a biggie; you can push it now if you want :)

ktoso reviewed (approved):
LGTM thanks! Thanks you made time for fixing it :)

dotta commented:
> Commit message is not a biggie; you can push it now if you want :)

Done :)

raboof commented on akka-testkit/src/test/scala/akka/testkit/TestEventListenerSpec.scala:
```diff
@@ -0,0 +1,60 @@
+    }
+  }
+
+  "The typed EventFilter" must {
```
Perhaps also add a testcase that `EventFilter[Error.NoCause]` has the 'old' behavior (only filtering errors that have no cause)?

dotta replied:
If this is the desired behavior, then the current fix should be reworked. I guess an explicit `Error.All` class should be introduced, and `ErrorFilter.error` should use that (as opposed to using `Error.NoCause` as it currently does). This boils down to decide if it's useful for the user to filter only errors that have no cause attached. So, is this what you'd like?

raboof replied:
Ah, indeed `EventFilter[Error.NoCause.type]` will now filter everything.

I think it would be a useful improvement to allow filtering on errors without a cause, but it's not essential for this PR.

raboof reviewed (commented):
Thanks for the contribution! We also need to update the scaladoc on `EventFilter.error()`, other than that LGTM! (one testcase that might be nice to add, but not essential)

raboof reviewed (changes requested):
Oh we should update the scaladoc before merging though

dotta commented:
> Oh we should update the scaladoc before merging though

@raboof Here is what the `EventFilter.error` scaladoc says:

```
/**
   * Create a filter for Error events which do not have a cause set (i.e. use
   * implicitly supplied Logging.Error.NoCause). See apply() for more details.
   */
```

Which is exactly what the code does. What would you like to change?

raboof commented:
The `which do not have a cause set` isn't really correct anymore, right? (it's true that this is achieved by passing in Logging.Error.NoCause, but that's more of an implementation detail now, right?)

dotta commented:
@raboof Agreed. Thanks for convincing me! :) Will make the change and force push

dotta commented:
@raboof Have another look. I've updated the scaladoc and also added an explicit test that showcase how the `EventFilter[Error.NoCause.type]` works atm (also added a link to the discussion to make it clear the current behavior could be improved).

raboof reviewed (approved):
Looks good, thanks, great contribution!

ktoso commented:
Great, thanks!
